modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1282 @@
+    const timeout = new timeval(4,0);
+    const connectedClient = connect(address, timeout);
+
+  :arg address: address to connect to
+  :type address: :type:`ipAddr`
+  :arg timeout: time to wait for connection establishment.
+  :type timeval: :type:`~Sys.timeval`
+  :return: connected socket.
+  :rtype: `tcpConn`
+  :throws SystemError: Upon failure to connect.
+*/
+proc connect(in address: ipAddr, in timeout = new timeval(-1,0)): tcpConn throws {
+  var family = address.family;
+  var socketFd = socket(family, SOCK_STREAM);
+  var err_out = sys_connect(socketFd, address._addressStorage);
+  if err_out != 0 && err_out != EINPROGRESS {
+    sys_close(socketFd);
+    throw SystemError.fromSyserr(err_out,"connect() failed");
+  }
+  if err_out == 0 {
+    return openfd(socketFd):tcpConn;
+  }
+  var localSync$: sync int = 0;
+  localSync$.readFE();
+  var writerEvent = event_new(event_loop_base, socketFd, EV_WRITE | EV_TIMEOUT | EV_ET, c_ptrTo(syncRWTCallback), c_ptrTo(localSync$):c_void_ptr);
+  defer {
+    event_del(writerEvent);
+    event_free(writerEvent);
+  }
+  if timeout.tv_sec == -1 {
+    err_out = event_add(writerEvent, nil);
+  }
+  else {
+    err_out = event_add(writerEvent, c_ptrTo(timeout));
+  }
+  if err_out != 0 {
+    throw new Error("connect() failed");
+  }
+  var retval = localSync$.readFE();
+  if retval & EV_TIMEOUT != 0 {
+    throw SystemError.fromSyserr(ETIMEDOUT);
+  }
+  err_out = sys_connect(socketFd, address._addressStorage);
+  if err_out != 0 {
+    sys_close(socketFd);
+    throw SystemError.fromSyserr(err_out,"connect() failed");
+  }
+  return openfd(socketFd):tcpConn;
+}
+
+/*
+  This overload of `connect` not only returns a :type:`tcpConn`
+  but also does DNS resolution for the provided `host`.
+  The `timeout` is tired for all resolved addresses and the first
+  successful one is returned back.
+
+  .. code-block:: Chapel
+
+    const timeout = new timeval(4,0);
+    const connectedClient = connect("google.com", "tcp", IPFamily.IPv4, timeout);
+
+  :arg host: host to connect to or resolve if not in standard ip notation
+  :type host: `string`
+  :arg service: service to connect to on resolved `host`
+  :type service: `string`
+  :arg family: type of socket family to connect to
+  :type family: :type:`IPFamily`
+  :arg timeout: time to wait for each possible connection.
+  :type timeval: :type:`~Sys.timeval`
+  :return: connected socket.
+  :rtype: `tcpConn`
+  :throws SystemError: Upon failure to resolve address or connect
+                        to any of the resolved address in given `timeout`.
+*/
+proc connect(in host: string, in service: string, family: IPFamily = IPFamily.IPUnspec, in timeout = new timeval(-1,0)): tcpConn throws {
+  var result:sys_addrinfo_ptr_t;
+  var hints = new sys_addrinfo_t();
+  hints.ai_family = family:c_int;
+  hints.ai_socktype = SOCK_STREAM;
+  var err = getaddrinfo(host.c_str(), service.c_str(), hints, result);
```

Comment:
the DNS lookup might itself take a long time and I would expect that to be subject to `timeout`. However, I don't know how to implement that. Can libevent do DNS lookup as an event?

Reply:
marked as future work